In [1]:
import sys
import platform
import numpy as np
import pandas as pd

print("Python      :", sys.version)
print("Implementación:", platform.python_implementation())
print("NumPy       :", np.__version__)
print("pandas      :", pd.__version__)

assert sys.version_info[:2] == (3, 12), "Esta práctica requiere Python 3.12"
print("\nEntorno verificado correctamente.")

Python      : 3.12.10 (tags/v3.12.10:0cc8128, Apr  8 2025, 12:21:36) [MSC v.1943 64 bit (AMD64)]
Implementación: CPython
NumPy       : 2.5.3
pandas      : 3.0.6

Entorno verificado correctamente.


## Ejercicio 1 — Catálogo de productos como lista

In [2]:
catalogo = [
    "Auriculares BT", 59.90,
    "Smartwatch S2", 149.00,
    "Tablet 10", 219.00,
    "Cargador USB-C", 18.50,
    "Cafetera Expres", 189.00,
    "Robot Aspirador", 279.00,
]

print(len(catalogo))

print("Número de productos:", len(catalogo) / 2)

#Saltamos de 2 en 2 por que la lista alterna entre nombres y precios
nombres = catalogo[0::2]
print(nombres)

precios = catalogo[1::2]
print(precios)

#Buscamos el nombre para encontrar su precio aunque cambie su posición.
posicion = catalogo.index("Robot Aspirador")
print("Precio del Robot Aspirador:", catalogo[posicion + 1])

ultimos_tres_nombres = nombres[-3:]
print(ultimos_tres_nombres)

precio_medio = round(sum(precios)/len(precios),2)
print(precio_medio)

12
Número de productos: 6.0
['Auriculares BT', 'Smartwatch S2', 'Tablet 10', 'Cargador USB-C', 'Cafetera Expres', 'Robot Aspirador']
[59.9, 149.0, 219.0, 18.5, 189.0, 279.0]
Precio del Robot Aspirador: 279.0
['Cargador USB-C', 'Cafetera Expres', 'Robot Aspirador']
152.4


## Ejercicio 2 — Inventario multialmacén con listas anidadas

In [3]:
inventario = [
    ["ALM-NORTE", "Robot Aspirador", 34, 201.50],
    ["ALM-NORTE", "Monitor 27", 58, 142.00],
    ["ALM-SUR",   "Silla Ergonomica", 12, 178.90],
    ["ALM-SUR",   "Robot Aspirador", 7, 201.50],
    ["ALM-ESTE",  "Cafetera Expres", 41, 131.20],
    ["ALM-ESTE",  "Monitor 27", 25, 142.00],
]

print(inventario[3][1])

valor_stock = []

# Calculamos el valor de cada fila porque un producto puede estar en varios almacenes.
for fila in inventario:
    producto = fila[1]
    valor = round(fila[2] * fila[3], 2)
    valor_stock.append([producto, valor])

print(valor_stock)

# Sumamos el valor de todas las filas para obtener el total del inventario.
total = 0

for elemento in valor_stock:
    total = total + elemento[1]

print("Valor total:", round(total, 2))

inventario_sur = inventario[2:4]
print(inventario_sur)

total_sur = 0

for fila in inventario_sur:
    total_sur = total_sur + fila[2] * fila[3]

print("Valor de ALM-SUR:", round(total_sur, 2))

Robot Aspirador
[['Robot Aspirador', 6851.0], ['Monitor 27', 8236.0], ['Silla Ergonomica', 2146.8], ['Robot Aspirador', 1410.5], ['Cafetera Expres', 5379.2], ['Monitor 27', 3550.0]]
Valor total: 27573.5
[['ALM-SUR', 'Silla Ergonomica', 12, 178.9], ['ALM-SUR', 'Robot Aspirador', 7, 201.5]]
Valor de ALM-SUR: 3557.3


### ¿Qué problema aparece si dos almacenes dejan de estar contiguos?
Que podemos dejar fuera parte de su inventario o podríamos influir el otro inventario.

### ¿Por qué el slicing no es una estrategia robusta de filtrado?
Porque seleciona posición y no comprueba el nombre del almacén

## Ejercicio 3 — Manipulación destructiva y copias de listas

In [4]:
precios_originales = [59.90, 149.00, 219.00, 18.50, 189.00, 279.00]

# 1. La asignación comparte la lista, por eso el cambio afecta a ambas variables.
precios_rebajados = precios_originales
precios_rebajados[0] = 49.90

print("Originales tras la asignación:", precios_originales)
print("Rebajados:", precios_rebajados)

# 2. Restauramos el precio antes de crear las copias independientes.
precios_originales[0] = 59.90

copia_lista = list(precios_originales)
copia_slicing = precios_originales[:]

copia_lista[0] = 39.90
copia_slicing[0] = 29.90

print("Originales sin cambios:", precios_originales)
print("Copia con list():", copia_lista)
print("Copia con slicing:", copia_slicing)

#3 Recuperamos el precio para aplicar la rebaja sobre los valores iniciales
copia_lista[0] = precios_originales[0]

for i in range(len(copia_lista)):
    copia_lista[i] = round(copia_lista[i] * 0.90, 2)

print("Copia con rebaja del 10 %:", copia_lista)

#4 Añadimos precios de ejemplo porque esta lista solo contiene precios
copia_lista.extend([45.00, 75.00])
copia_lista.insert(2, 25.00)

print("Tras añadir los tres productos:", copia_lista)

#5 Guardamos el último precio para conservar el dato que retiramos.
del copia_lista[4]
descartado = copia_lista.pop()

print("Tras eliminar los dos elementos:", copia_lista)
print("Precio descartado con pop():", descartado)

#6 Comprobamos que las copias son objetos distintos del original
print("ID del original:", id(precios_originales))
print("ID de la copia con list():", id(copia_lista))
print("ID de la copia con slicing:", id(copia_slicing))

print("Original distinto de copia_lista:",
        id(precios_originales) != id(copia_lista))
print("Original distinto de copia_slicing:",
        id(precios_originales) != id(copia_slicing))
print("Copias distintas entre sí:",
        id(copia_lista) != id(copia_slicing))

print("Originales al terminar:", precios_originales)

Originales tras la asignación: [49.9, 149.0, 219.0, 18.5, 189.0, 279.0]
Rebajados: [49.9, 149.0, 219.0, 18.5, 189.0, 279.0]
Originales sin cambios: [59.9, 149.0, 219.0, 18.5, 189.0, 279.0]
Copia con list(): [39.9, 149.0, 219.0, 18.5, 189.0, 279.0]
Copia con slicing: [29.9, 149.0, 219.0, 18.5, 189.0, 279.0]
Copia con rebaja del 10 %: [53.91, 134.1, 197.1, 16.65, 170.1, 251.1]
Tras añadir los tres productos: [53.91, 134.1, 25.0, 197.1, 16.65, 170.1, 251.1, 45.0, 75.0]
Tras eliminar los dos elementos: [53.91, 134.1, 25.0, 197.1, 170.1, 251.1, 45.0]
Precio descartado con pop(): 75.0
ID del original: 1946386914752
ID de la copia con list(): 1946386921600
ID de la copia con slicing: 1946386915200
Original distinto de copia_lista: True
Original distinto de copia_slicing: True
Copias distintas entre sí: True
Originales al terminar: [59.9, 149.0, 219.0, 18.5, 189.0, 279.0]


Al asignar una lista a otra variable, ambas apuntan a la misma lista y el cambio aparece en las dos. Con `list()` y `[:]` creamos copias independientes y comprobamos que sus cambios no afectan al original.

## Ejercicio 4 — Cola de pedidos pendientes

In [5]:
cola = ["PED-10021", "PED-10007", "PED-10044", "PED-10012", "PED-10033", "PED-10008"]

# Recorremos la cola desde el final sin cambiar la lista original.
cola_invertida = cola[::-1]
print("Cola invertida:", cola_invertida)

copia_original = cola[:]
cola_ordenada = sorted(cola)

print("Orden alfabético:", cola_ordenada)
print("Cola original:", cola)
print("La original sigue intacta:", cola == copia_original)

# Ahora sí cambiamos el orden de la propia cola.
cola.sort(reverse=True)
print("Cola ordenada de mayor a menor:", cola)

# Al retirar el primero, el siguiente pasa a ocupar la posición 0.
atendidos = []
atendidos.append(cola.pop(0))
atendidos.append(cola.pop(0))

print("Pedidos atendidos:", atendidos)

cola.insert(0, "PED-99999")
print("Cola con el pedido urgente:", cola)

# Incluimos el nuevo pedido urgente en el total de trabajo.
numero_atendidos = len(atendidos)
numero_pendientes = len(cola)
total_pedidos = numero_atendidos + numero_pendientes
avance = round(numero_atendidos / total_pedidos * 100, 1)

print("Número de pedidos atendidos:", numero_atendidos)
print("Número de pedidos pendientes:", numero_pendientes)
print("Porcentaje de avance:", avance, "%")


Cola invertida: ['PED-10008', 'PED-10033', 'PED-10012', 'PED-10044', 'PED-10007', 'PED-10021']
Orden alfabético: ['PED-10007', 'PED-10008', 'PED-10012', 'PED-10021', 'PED-10033', 'PED-10044']
Cola original: ['PED-10021', 'PED-10007', 'PED-10044', 'PED-10012', 'PED-10033', 'PED-10008']
La original sigue intacta: True
Cola ordenada de mayor a menor: ['PED-10044', 'PED-10033', 'PED-10021', 'PED-10012', 'PED-10008', 'PED-10007']
Pedidos atendidos: ['PED-10044', 'PED-10033']
Cola con el pedido urgente: ['PED-99999', 'PED-10021', 'PED-10012', 'PED-10008', 'PED-10007']
Número de pedidos atendidos: 2
Número de pedidos pendientes: 5
Porcentaje de avance: 28.6 %


## Ejercicio 5 — Funciones propias para el cálculo de importes

In [6]:
def calcular_importe(unidades, precio_unitario, descuento_pct=0):
    """Calcula el importe con descuento y lo redondea a dos decimales."""
    bruto = unidades * precio_unitario
    descuento = bruto * descuento_pct / 100
    return round(bruto - descuento, 2)


def resumen_pedido(unidades, precio_unitario, descuento_pct=0):
    """Devuelve una tupla con el importe bruto, el ahorro y el importe neto."""
    importe_bruto = unidades * precio_unitario
    ahorro = importe_bruto * descuento_pct / 100
    # Reutilizamos la función para aplicar el mismo cálculo del neto.
    importe_neto = calcular_importe(unidades, precio_unitario, descuento_pct)
    return round(importe_bruto, 2), round(ahorro, 2), importe_neto


print("Importe de comprobación:", calcular_importe(3, 59.90, 10))
print("Importe sin descuento:", calcular_importe(3, 59.90))

# Indicamos los nombres de los parámetros y recogemos los tres resultados.
bruto, ahorro, neto = resumen_pedido(
    unidades=3, precio_unitario=59.90, descuento_pct=10
)
print("Importe bruto:", bruto)
print("Ahorro:", ahorro)
print("Importe neto:", neto)

help(calcular_importe)

pedidos = [
    (3, 59.90, 10),
    (1, 279.00, 0),
    (5, 18.50, 20),
    (2, 149.00, 5),
    (4, 89.90, 15),
]

importes = []
for unidades, precio, descuento in pedidos:
    importe = calcular_importe(unidades, precio, descuento)
    importes.append(importe)

print("Importes:", importes)
print("Importe máximo:", max(importes))
print("Importe mínimo:", min(importes))
print("Importes ordenados:", sorted(importes))
print("Número de pedidos:", len(importes))


Importe de comprobación: 161.73
Importe sin descuento: 179.7
Importe bruto: 179.7
Ahorro: 17.97
Importe neto: 161.73
Help on function calcular_importe in module __main__:

calcular_importe(unidades, precio_unitario, descuento_pct=0)
    Calcula el importe con descuento y lo redondea a dos decimales.

Importes: [161.73, 279.0, 74.0, 283.1, 305.66]
Importe máximo: 305.66
Importe mínimo: 74.0
Importes ordenados: [74.0, 161.73, 279.0, 283.1, 305.66]
Número de pedidos: 5


## Ejercicio 6 — Normalización de códigos SKU con métodos

In [7]:
skus_crudos = [
    "  elc-0012-es ", "HOG-0045-ES", "dep-0003-pt  ",
    "ofi-0021-es", "  ELC-0012-ES", "hog-0099-fr ",
]

# Unificamos el formato para poder comparar códigos que representan lo mismo.
skus_limpios = []
for sku in skus_crudos:
    skus_limpios.append(sku.strip().upper())

print("Códigos limpios:",skus_limpios)
print("Veces que aparece ELC-0012-ES:", skus_limpios.count("ELC-0012-ES"))
print("Posición de OFI-0021-ES:", skus_limpios.index("OFI-0021-ES"))

partes_skus = []
for sku in skus_limpios:
    partes_skus.append(sku.split("-"))

print("Familia, número y país:", partes_skus)

skus_es = []
for sku in skus_limpios:
    partes = sku.split("-")
    # Comprobamos el país completo, que ocupa la tercera parte del código.
    if partes[2] == "ES":
        skus_es.append(sku.replace("-","."))

print("Códigos de España:", skus_es)
cadena_skus = " | ".join(skus_es)
print("Cadena final:",cadena_skus)


Códigos limpios: ['ELC-0012-ES', 'HOG-0045-ES', 'DEP-0003-PT', 'OFI-0021-ES', 'ELC-0012-ES', 'HOG-0099-FR']
Veces que aparece ELC-0012-ES: 2
Posición de OFI-0021-ES: 3
Familia, número y país: [['ELC', '0012', 'ES'], ['HOG', '0045', 'ES'], ['DEP', '0003', 'PT'], ['OFI', '0021', 'ES'], ['ELC', '0012', 'ES'], ['HOG', '0099', 'FR']]
Códigos de España: ['ELC.0012.ES', 'HOG.0045.ES', 'OFI.0021.ES', 'ELC.0012.ES']
Cadena final: ELC.0012.ES | HOG.0045.ES | OFI.0021.ES | ELC.0012.ES


## Ejercicio 7 — Paquetes, importaciones y módulo propio

In [8]:
import math

print("Raíz cuadrada de 2809:", math.sqrt(2809))

from math import pi, ceil

#Redondeamos hacia arriba porque las unidades restantes necesitan otro palé.
pales = ceil(1375 / 24)
print("Palés necesarios:",pales)

import numpy as np
import pandas as pd

print("NumPy:", np.__version__)
print("pandas:", pd.__version__)

import random

random.seed(42)
valoraciones = [random.randint(1, 5) for _ in range(10)]
print(valoraciones)

# El módulo está en src, una carpeta hermana de notebooks.
import sys
sys.path.append("../src")

from utilidades import calcular_importe, clasificar_ticket

print(calcular_importe(4, 89.90, 15))
print(clasificar_ticket(305.66))


Raíz cuadrada de 2809: 53.0
Palés necesarios: 58
NumPy: 2.5.3
pandas: 3.0.6
[1, 1, 3, 2, 2, 2, 1, 5, 1, 5]
305.66
Medio


Al fijar la misma semilla, obtenemos la misma secuencia de valoraciones al ejecutar el código en el mismo entorno.


## Ejercicio 8 — Del cálculo con listas al cálculo vectorizado

In [9]:
import numpy as np

unidades = [3, 1, 5, 2, 4, 6, 2, 8, 3, 1]
precios = [59.90, 279.00, 18.50, 149.00, 89.90, 27.50, 219.00, 14.90, 64.00, 239.00]

# Mostramos el error esperado sin detener los siguientes apartados.
try:
    print(unidades * precios)
except TypeError as error:
    print("Error al multiplicar listas:", error)

np_unidades = np.array(unidades)
np_precios = np.array(precios)
print("Unidades: dtype y shape:", np_unidades.dtype, np_unidades.shape)
print("Precios: dtype y shape:", np_precios.dtype, np_precios.shape)

importes = np_unidades * np_precios
print("Importes:",importes)
print("Suma total:", round(importes.sum(), 2))

# Conservamos importes sin recargo porque se utiliza en el siguiente ejercicio.
importes_con_recargo = importes * 1.04
print("Importes con recargo del 4 %:", importes_con_recargo)
print("Precios superiores a 100:", np.count_nonzero(np_precios > 100))

mixto = np.array([1, "dos", 3.0, True])
print("Array mixto:", mixto)
print("Tipo del array mixto:",mixto.dtype)


def calcular_con_bucle():
    resultado = []
    for i in range(len(unidades)):
        resultado.append(unidades[i] * precios[i])
    return resultado


Error al multiplicar listas: can't multiply sequence by non-int of type 'list'
Unidades: dtype y shape: int64 (10,)
Precios: dtype y shape: float64 (10,)
Importes: [179.7 279.   92.5 298.  359.6 165.  438.  119.2 192.  239. ]
Suma total: 2362.0
Importes con recargo del 4 %: [186.888 290.16   96.2   309.92  373.984 171.6   455.52  123.968 199.68
 248.56 ]
Precios superiores a 100: 4
Array mixto: ['1' 'dos' '3.0' 'True']
Tipo del array mixto: <U32


In [10]:
%timeit calcular_con_bucle()
%timeit np_unidades * np_precios


728 ns ± 148 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
967 ns ± 54.3 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)



Al multiplicar las dos listas obtenemos un TypeError porque Python no permite multiplicarlas elemento a elemento.
El array mixto tiene dtype <U32: convierte los valores a texto para mantener un mismo tipo de dato.
En esta ejecución, el bucle tarda 728 nanosegundos y NumPy 967 nanosegundos, aunque esta prueba con diez elementos no permite concluir que el bucle sea siempre más rápido.


## Ejercicio 9 — Subsetting con máscaras booleanas

In [11]:
# Reutilizamos los arrays del ejercicio 8, con los importes sin recargo.
caros = np_precios > 100
print("Máscara de precios caros:", caros)
print("Precios superiores a 100:",np_precios[caros])
print("Número de productos caros:", caros.sum())

# Las posiciones coinciden en ambos arrays, por eso sirve la misma máscara.
importes_caros = importes[caros]
print("Importes de esas líneas:", importes_caros)
print("Suma de sus importes:", round(importes_caros.sum(), 2))
print("Posiciones:", np.where(caros))

varias = np_unidades >= 2
print("Máscara de dos o más unidades:",varias)

seleccion = (np_precios > 100) & (np_unidades >= 2)
print("Importes con ambas condiciones:", importes[seleccion])


Máscara de precios caros: [False  True False  True False False  True False False  True]
Precios superiores a 100: [279. 149. 219. 239.]
Número de productos caros: 4
Importes de esas líneas: [279. 298. 438. 239.]
Suma de sus importes: 1254.0
Posiciones: (array([1, 3, 6, 9]),)
Máscara de dos o más unidades: [ True False  True  True  True  True  True  True  True False]
Importes con ambas condiciones: [298. 438.]


Al sumar booleanos, True cuenta como 1 y False como 0, por eso obtenemos el número de productos que cumplen la condición.

## Ejercicio 10 — Matriz de ventas por trimestre y región

In [12]:
import numpy as np

# Filas: T1, T2, T3. Columnas: Norte, Sur, Este, Oeste, Centro.
ventas_2d = np.array([
    [15200.50, 11800.25, 12950.00, 9870.40, 13400.10],
    [16840.75, 12310.60, 13480.90, 11020.35, 14100.80],
    [12535.72, 10663.06, 11221.59, 12673.08, 12334.38],
])

print("Forma:", ventas_2d.shape)
print("Dimensiones:",ventas_2d.ndim)
print("Número de valores:", ventas_2d.size)

print("Segundo trimestre:", ventas_2d[1, :])
print("Región Centro:", ventas_2d[:, 4])
print("T3 en la región Este:", ventas_2d[2, 2])
print("T1 y T2, tres primeras regiones:")
print(ventas_2d[:2, :3])

# Sumamos las columnas de cada fila para obtener el total trimestral.
totales_trimestre = ventas_2d.sum(axis=1)
totales_region = ventas_2d.sum(axis=0)
print("Totales por trimestre:", np.round(totales_trimestre, 2))
print("Totales por región:", np.round(totales_region, 2))

# Añadimos la previsión aparte para conservar las ventas reales de los tres trimestres.
prevision = ventas_2d[2] * 1.06
ventas_con_prevision = np.vstack((ventas_2d, prevision))
print("Ventas con la cuarta fila de previsión:")
print(np.round(ventas_con_prevision, 2))

regiones = ["Norte", "Sur", "Este", "Oeste", "Centro"]
indice_mayor = np.argmax(totales_region)
print("Región con mayor facturación:",regiones[indice_mayor])
print("Facturación acumulada:", round(totales_region[indice_mayor], 2))


Forma: (3, 5)
Dimensiones: 2
Número de valores: 15
Segundo trimestre: [16840.75 12310.6  13480.9  11020.35 14100.8 ]
Región Centro: [13400.1  14100.8  12334.38]
T3 en la región Este: 11221.59
T1 y T2, tres primeras regiones:
[[15200.5  11800.25 12950.  ]
 [16840.75 12310.6  13480.9 ]]
Totales por trimestre: [63221.25 67753.4  59427.83]
Totales por región: [44576.97 34773.91 37652.49 33563.83 39835.28]
Ventas con la cuarta fila de previsión:
[[15200.5  11800.25 12950.    9870.4  13400.1 ]
 [16840.75 12310.6  13480.9  11020.35 14100.8 ]
 [12535.72 10663.06 11221.59 12673.08 12334.38]
 [13287.86 11302.84 11894.89 13433.46 13074.44]]
Región con mayor facturación: Norte
Facturación acumulada: 44576.97


shape devuelve (3, 5), porque tenemos tres trimestres y cinco regiones. ndim devuelve 2, porque organizamos las ventas en filas y columnas. size devuelve 15, que es el número total de valores de ventas.

## Ejercicio 11 — Estadística descriptiva sobre telemetría simulada

In [13]:
import numpy as np

# Reiniciamos el generador con la misma semilla para repetir estos datos.
rng = np.random.default_rng(2024)
temperatura = rng.normal(68, 6.5, 500)
vibracion = np.abs(rng.normal(2.4, 0.8, 500))

media = np.mean(temperatura)
mediana = np.median(temperatura)
desviacion = np.std(temperatura)

print("Media:", media)
print("Mediana:",mediana)
print("Diferencia entre media y mediana:", media - mediana)
print("Desviación típica:", desviacion)
print("Temperatura mínima:", np.min(temperatura))
print("Temperatura máxima:", np.max(temperatura))

percentiles = np.percentile(vibracion, [25, 50, 75, 95])
print("Percentiles 25, 50, 75 y 95 de vibración:", percentiles)

mediciones = np.column_stack((temperatura, vibracion))
print("Forma de la matriz:",mediciones.shape)

correlacion = np.corrcoef(temperatura, vibracion)
print("Matriz de correlación:")
print(correlacion)
print("Correlación entre temperatura y vibración:", correlacion[0, 1])

# Usamos los valores sin redondear para decidir qué lecturas superan el umbral.
umbral = media + 2 * desviacion
superan = np.count_nonzero(temperatura > umbral)
porcentaje = superan / temperatura.size * 100

print("Umbral de temperatura:", umbral)
print("Lecturas que superan el umbral:",superan)
print("Porcentaje de lecturas:", round(porcentaje, 2), "%")


Media: 68.03192414779825
Mediana: 68.13342844453155
Diferencia entre media y mediana: -0.10150429673329597
Desviación típica: 6.501603731938141
Temperatura mínima: 49.014430627351686
Temperatura máxima: 88.2903599137993
Percentiles 25, 50, 75 y 95 de vibración: [1.89488326 2.39381414 2.92451892 3.72951895]
Forma de la matriz: (500, 2)
Matriz de correlación:
[[1.         0.00673339]
 [0.00673339 1.        ]]
Correlación entre temperatura y vibración: 0.006733392213856878
Umbral de temperatura: 81.03513161167453
Lecturas que superan el umbral: 11
Porcentaje de lecturas: 2.2 %


Al fijar la semilla 2024, generamos los mismos datos al repetir el código en el mismo entorno. Así, otra persona puede reproducir el análisis y comprobar los resultados.

La media (68,03) y la mediana (68,13) son muy próximas, lo que sugiere que las temperaturas están bastante equilibradas alrededor del centro.
La correlación es aproximadamente 0,0067, muy cercana a cero, por lo que no observamos una relación lineal apreciable entre temperatura y vibración.

## Ejercicio 12 — Diccionarios simples y anidados

In [14]:
tarifas_envio = {
    "Norte": 4.95,
    "Sur": 5.50,
    "Este": 5.20,
    "Oeste": 6.10,
    "Centro": 3.90,
}

print("Tarifa de Este:", tarifas_envio["Este"])
tarifas_envio["Insular"] = 9.80
tarifas_envio["Oeste"] = 5.95

del tarifas_envio["Sur"]
print("¿Sigue Sur en las tarifas?", "Sur" in tarifas_envio)

for region, tarifa in tarifas_envio.items():
    print(f"{region} -> {tarifa:.2f} €")

red_logistica = {
    "Norte": {"almacen": "ALM-NORTE", "tarifa": 4.95, "entrega_h": 24},
    "Este": {"almacen": "ALM-ESTE", "tarifa": 5.20, "entrega_h": 48},
    "Centro": {"almacen": "ALM-CENTRO", "tarifa": 3.90, "entrega_h": 24},
}

print("Horas de entrega en Este:",red_logistica["Este"]["entrega_h"])

# Elegimos 48 horas como ejemplo para Oeste, ya que no se fija ese plazo.
red_logistica["Oeste"] = {
    "almacen": "ALM-OESTE",
    "tarifa": tarifas_envio["Oeste"],
    "entrega_h": 48,
}

# El primer bucle recorre regiones y el segundo, los datos de cada una.
for region, datos in red_logistica.items():
    print(f"\nRegión: {region}")
    for campo, valor in datos.items():
        print(f"  {campo}: {valor}")


Tarifa de Este: 5.2
¿Sigue Sur en las tarifas? False
Norte -> 4.95 €
Este -> 5.20 €
Oeste -> 5.95 €
Centro -> 3.90 €
Insular -> 9.80 €
Horas de entrega en Este: 48

Región: Norte
  almacen: ALM-NORTE
  tarifa: 4.95
  entrega_h: 24

Región: Este
  almacen: ALM-ESTE
  tarifa: 5.2
  entrega_h: 48

Región: Centro
  almacen: ALM-CENTRO
  tarifa: 3.9
  entrega_h: 24

Región: Oeste
  almacen: ALM-OESTE
  tarifa: 5.95
  entrega_h: 48


Usamos un diccionario cuando queremos acceder a un dato por una clave sin depender de su posición. La clave identifica qué representa el dato, como una región, mientras que en una lista accedemos normalmente mediante un índice.

## Ejercicio 13 — De diccionario a DataFrame y lectura de CSV

In [15]:
import pandas as pd

almacenes = {
    "codigo": ["ALM-NORTE", "ALM-SUR", "ALM-ESTE", "ALM-OESTE", "ALM-CENTRO"],
    "ciudad": ["Bilbao", "Sevilla", "Valencia", "Vigo", "Madrid"],
    "m2": [4200, 3100, 3800, 2600, 6500],
    "operarios": [48, 31, 39, 22, 74],
}

almacenes = pd.DataFrame(almacenes)
print("Almacenes con índice numérico:")
print(almacenes)

# El código identifica cada almacén mejor que una posición numérica.
almacenes = almacenes.set_index("codigo")
print("\nAlmacenes con código como índice:")
print(almacenes)

# Las rutas parten de notebooks y conservamos los CSV sin modificar.
ventas = pd.read_csv("../data/ventas_retail.csv")
empleados = pd.read_csv("../data/empleados.csv", index_col="id_empleado")
sensores = pd.read_csv("../data/sensores_planta.csv")

print("\nPrimeras ocho ventas:")
print(ventas.head(8))
print("\nÚltimas tres ventas:")
print(ventas.tail(3))
print("Forma de ventas:",ventas.shape)

print("\nTipos de columnas y datos disponibles:")
ventas.info()

print("\nValores ausentes por columna:")
print(ventas.isna().sum())

print("\nForma de los tres DataFrames:")
print("Ventas:", ventas.shape)
print("Empleados:",empleados.shape)
print("Sensores:", sensores.shape)


Almacenes con índice numérico:
       codigo    ciudad    m2  operarios
0   ALM-NORTE    Bilbao  4200         48
1     ALM-SUR   Sevilla  3100         31
2    ALM-ESTE  Valencia  3800         39
3   ALM-OESTE      Vigo  2600         22
4  ALM-CENTRO    Madrid  6500         74

Almacenes con código como índice:
              ciudad    m2  operarios
codigo                               
ALM-NORTE     Bilbao  4200         48
ALM-SUR      Sevilla  3100         31
ALM-ESTE    Valencia  3800         39
ALM-OESTE       Vigo  2600         22
ALM-CENTRO    Madrid  6500         74

Primeras ocho ventas:
   id_pedido       fecha  region    categoria          producto  unidades  \
0  PED-10153  2025-01-01  Centro  Electronica        Tablet 10"         2   
1  PED-10030  2025-01-02     Sur      Deporte   Zapatilla Trail         4   
2  PED-10343  2025-01-02   Oeste      Oficina  Teclado Mecanico         3   
3  PED-10298  2025-01-02   Oeste        Hogar   Robot Aspirador         6   
4  PED-10144  

Las columnas numéricas son unidades, precio_unitario, descuento_pct y satisfaccion. Las columnas de texto son id_pedido, fecha, region, categoria, producto y canal; fecha se ha leído como texto.
Los valores ausentes se concentran en canal, con 5, y satisfaccion, con 9.

## Ejercicio 14 — Selección con corchetes, loc e iloc

In [16]:
salario_serie = empleados["salario_base"]
salario_tabla = empleados[["salario_base"]]
print("Tipo con un par de corchetes:", type(salario_serie))
print("Tipo con doble corchete:",type(salario_tabla))
print(salario_serie)
print(salario_tabla)

print("\nDepartamento, ciudad y salario:")
print(empleados[["departamento", "ciudad", "salario_base"]])

# El límite final no se incluye: usamos 16 para abarcar la posición 15.
print("\nFilas en las posiciones 10 a 15:")
print(empleados[10:16])

print("\nFicha de EMP-1042:")
print(empleados.loc["EMP-1042"])

print("\nDepartamento y modalidad de los tres empleados:")
print(empleados.loc[
    ["EMP-1005", "EMP-1020", "EMP-1099"],
    ["departamento", "modalidad"]
])

print("\nTres primeras filas, columnas en posiciones 2 y 5:")
print(empleados.iloc[:3, [2, 5]])

# Los dos puntos conservan todas las filas; los nombres eligen las columnas.
print("\nNombre y porcentaje de bonus:")
print(empleados.loc[:, ["nombre", "bonus_pct"]])


Tipo con un par de corchetes: <class 'pandas.Series'>
Tipo con doble corchete: <class 'pandas.DataFrame'>
id_empleado
EMP-1001    43297
EMP-1002    32554
EMP-1003    47738
EMP-1004    39733
EMP-1005    44990
            ...  
EMP-1176    37301
EMP-1177    44604
EMP-1178    46786
EMP-1179    35736
EMP-1180    46315
Name: salario_base, Length: 180, dtype: int64
             salario_base
id_empleado              
EMP-1001            43297
EMP-1002            32554
EMP-1003            47738
EMP-1004            39733
EMP-1005            44990
...                   ...
EMP-1176            37301
EMP-1177            44604
EMP-1178            46786
EMP-1179            35736
EMP-1180            46315

[180 rows x 1 columns]

Departamento, ciudad y salario:
            departamento     ciudad  salario_base
id_empleado                                      
EMP-1001      Ingenieria   Valencia         43297
EMP-1002         Soporte   Valencia         32554
EMP-1003     Operaciones   Valencia        

| Forma | Selección | Ejemplo |
| --- | --- | --- |
| Corchetes `[]` | Columnas por nombre. Un slice entero selecciona filas por posición y excluye el final. | `empleados["salario_base"]` es una Serie. `empleados[["salario_base"]]` es un DataFrame. `empleados[10:16]` toma las posiciones 10 a 15. |
| `loc` | Filas y columnas por etiqueta. En un slice de etiquetas se incluyen ambos extremos cuando están presentes. | `empleados.loc["EMP-1042"]` busca ese identificador, no la posición 1042. |
| `iloc` | Filas y columnas por posición desde cero. El final del slice queda excluido. | `empleados.iloc[:3, [2, 5]]` toma las filas 0, 1 y 2 y las columnas 2 y 5. |


## Ejercicio 15 — Columnas calculadas, agregación y exportación

In [17]:
from decimal import Decimal, ROUND_HALF_EVEN

ventas["importe_bruto"] = ventas["unidades"] * ventas["precio_unitario"]

# Calculamos con decimales exactos para no desplazar los importes de medio céntimo.
netos = []
for unidades, precio, descuento in zip(
    ventas["unidades"], ventas["precio_unitario"], ventas["descuento_pct"]
):
    bruto = Decimal(str(unidades)) * Decimal(str(precio))
    neto = bruto * (1 - Decimal(str(descuento)) / 100)
    netos.append(float(neto.quantize(Decimal("0.01"), rounding=ROUND_HALF_EVEN)))

ventas["importe_neto"] = netos
ventas["ticket_medio"] = ventas["importe_neto"] / ventas["unidades"]

print("Primeras ventas con los importes calculados:")
print(ventas[["id_pedido", "importe_bruto", "importe_neto", "ticket_medio"]].head())

facturacion_region = ventas.groupby("region")["importe_neto"].sum().sort_values(ascending=False)
print("\nFacturación por región:")
print(facturacion_region)
print("Facturación neta total:", round(ventas["importe_neto"].sum(), 2))

# Conservamos también las ventas cuyo canal todavía está vacío.
resumen_categoria_canal = ventas.groupby(["categoria", "canal"], dropna=False).agg({
    "importe_neto": "mean",
    "unidades": "sum",
})
print("\nImporte medio y unidades por categoría y canal:")
print(resumen_categoria_canal)

top_productos = ventas.groupby("producto")["importe_neto"].sum().nlargest(5)
print("\nCinco productos con mayor facturación:")
print(top_productos)

# Rellenamos los ausentes en memoria sin alterar los CSV originales.
mediana_satisfaccion = ventas["satisfaccion"].median()
ventas["satisfaccion"] = ventas["satisfaccion"].fillna(mediana_satisfaccion)
ventas["canal"] = ventas["canal"].fillna("Desconocido")
print("\nMediana usada para satisfacción:",mediana_satisfaccion)
print("Ausentes después del relleno:")
print(ventas[["satisfaccion", "canal"]].isna().sum())

# Conservamos el índice para que cada importe siga asociado a su región.
facturacion_region.to_csv("../outputs/facturacion_por_region.csv", encoding="utf-8")
print("\nArchivo guardado: ../outputs/facturacion_por_region.csv")


Primeras ventas con los importes calculados:
   id_pedido  importe_bruto  importe_neto  ticket_medio
0  PED-10153          438.0        350.40        175.20
1  PED-10030          476.0        476.00        119.00
2  PED-10343          269.7        242.73         80.91
3  PED-10298         1674.0       1422.90        237.15
4  PED-10144          320.0        256.00         51.20

Facturación por región:
region
Norte     44576.99
Centro    39835.28
Este      37652.49
Sur       34773.90
Oeste     33563.82
Name: importe_neto, dtype: float64
Facturación neta total: 190402.48

Importe medio y unidades por categoría y canal:
                         importe_neto  unidades
categoria   canal                              
Deporte     Marketplace    311.646000       111
            Online         292.234583       225
            Tienda         189.056667       157
            NaN             32.000000         1
Electronica Marketplace    427.720870        95
            Online         426.703208 

Calculamos los importes con `Decimal` y redondeamos cada pedido a dos decimales con `ROUND_HALF_EVEN`. En los empates exactos de medio céntimo elegimos el céntimo par. Después sumamos los importes redondeados. Por eso Norte suma 44.576,99 €, mientras que al sumar sin redondear cada pedido y redondear solo el total obtenemos 44.576,97 €.


## Ejercicio 16 — Operadores de comparación y booleanos

In [18]:
import numpy as np

print(2 < 3)
print("Norte" == "norte")
print(True + True + False)
print(3 != 3.0)
print("Zapatilla" > "Mochila")

unidades = 6
precio = 149.0
descuento = 15

pedido_cumple = unidades > 4 and descuento >= 10
precio_fuera_intervalo = precio < 50 or precio > 200
print("Más de cuatro unidades y descuento de al menos 10 %:", pedido_cumple)
print("Precio inferior a 50 o superior a 200:",precio_fuera_intervalo)
print("El pedido no cumple ambas condiciones:", not pedido_cumple)

temp = np.array([64.2, 88.1, 71.0, 92.5, 69.8])
vib = np.array([1.9, 3.4, 2.1, 4.0, 2.2])

# Conservamos el error solicitado sin detener la ejecución del notebook.
try:
    print(temp > 70 and vib > 3)
except ValueError as error:
    print("Error con and:", error)

resultado_logical = np.logical_and(temp > 70, vib > 3)
resultado_operador = (temp > 70) & (vib > 3)
print("Con logical_and:", resultado_logical)
print("Con &:",resultado_operador)
print("Resultados idénticos:", np.array_equal(resultado_logical, resultado_operador))


True
False
2
False
True
Más de cuatro unidades y descuento de al menos 10 %: True
Precio inferior a 50 o superior a 200: False
El pedido no cumple ambas condiciones: False
Error con and: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
Con logical_and: [False  True False  True False]
Con &: [False  True False  True False]
Resultados idénticos: True


`2 < 3` da `True` porque 2 es menor que 3.

`"Norte" == "norte"` da `False` porque distingue mayúsculas y minúsculas.

`True + True + False` da `2` porque `True` vale 1 y `False` vale 0.

`3 != 3.0` da `False` porque ambos tienen el mismo valor.

`"Zapatilla" > "Mochila"` da `True` porque la Z va después de la M.

`not` da `False` porque el pedido sí supera cuatro unidades y tiene al menos un 10 % de descuento.

`and` produce un `ValueError` porque intenta interpretar todo el array como un único verdadero o falso.


## Ejercicio 17 — if / elif / else aplicado a mantenimiento predictivo

In [19]:
def nivel_riesgo(temperatura):
    if temperatura > 85:
        return "CRÍTICO"
    elif temperatura > 76:
        return "ALTO"
    elif temperatura > 70:
        return "MEDIO"
    else:
        return "BAJO"


for valor in [92.5, 80.0, 72.3, 64.1]:
    print(valor, nivel_riesgo(valor))

# Pasamos la función sin llamarla para que apply la use con cada temperatura.
sensores["riesgo"] = sensores["temperatura_c"].apply(nivel_riesgo)
print(sensores[["id_maquina", "temperatura_c", "riesgo"]].head())

print("\nLecturas por nivel de riesgo:")
print(sensores["riesgo"].value_counts())

criticas = sensores[sensores["riesgo"] == "CRÍTICO"]
criticas_por_maquina = criticas["id_maquina"].value_counts()
print("\nLecturas críticas por máquina:")
print(criticas_por_maquina)

# Mostramos todas las máquinas empatadas en el máximo, si hay más de una.
if not criticas_por_maquina.empty:
    maximo = criticas_por_maquina.max()
    print("\nMáquinas con más lecturas críticas:")
    print(criticas_por_maquina[criticas_por_maquina == maximo])
else:
    print("No hay lecturas con riesgo CRÍTICO.")

# El estado del CSV es distinto del riesgo que calculamos por temperatura.
print("\nLecturas con estado ALARMA:", (sensores["estado"] == "ALARMA").sum())
print("Lecturas con estado AVISO:",(sensores["estado"] == "AVISO").sum())


92.5 CRÍTICO
80.0 ALTO
72.3 MEDIO
64.1 BAJO
  id_maquina  temperatura_c riesgo
0     MAQ-01          58.92   BAJO
1     MAQ-03          71.72  MEDIO
2     MAQ-02          67.93   BAJO
3     MAQ-03          73.10  MEDIO
4     MAQ-03          69.37   BAJO

Lecturas por nivel de riesgo:
riesgo
BAJO       347
MEDIO      161
ALTO        72
CRÍTICO     20
Name: count, dtype: int64

Lecturas críticas por máquina:
id_maquina
MAQ-04    6
MAQ-02    5
MAQ-01    5
MAQ-03    4
Name: count, dtype: int64

Máquinas con más lecturas críticas:
id_maquina
MAQ-04    6
Name: count, dtype: int64

Lecturas con estado ALARMA: 20
Lecturas con estado AVISO: 72


Priorizamos la revisión de MAQ-04 porque acumula 6 lecturas críticas, más que las otras máquinas.  
Revisamos también las lecturas críticas de las demás máquinas y vigilamos las 72 de riesgo alto para detectar posibles empeoramientos.


## Ejercicio 18 — Filtrado avanzado de DataFrames

In [20]:
online_mayor_500 = ventas[
    (ventas["canal"] == "Online") & (ventas["importe_neto"] > 500)
]
print("Pedidos Online de más de 500 euros:")
print(online_mayor_500)

norte_centro = ventas[ventas["region"].isin(["Norte", "Centro"])]
print("\nPedidos de Norte o Centro:")
print(norte_centro)

# La disyunción representa aquí la misma selección que isin.
mascara_regiones = (ventas["region"] == "Norte") | (ventas["region"] == "Centro")
print("isin y | coinciden:", norte_centro.equals(ventas[mascara_regiones]))

con_descuento = ventas[ventas["descuento_pct"].between(10, 20)]
print("\nDescuentos entre 10 y 20 incluidos:")
print(con_descuento)

# Exigimos las tres condiciones a la vez para respetar la selección descrita.
electronica = ventas[
    (ventas["categoria"] == "Electronica")
    & (ventas["unidades"] > 2)
    & (ventas["descuento_pct"] != 0)
]
print("\nElectrónica, más de dos unidades y descuento distinto de cero:")
print(electronica)

sin_oficina = ventas[~(ventas["categoria"] == "Oficina")]
print("\nPedidos que no son de Oficina:")
print(sin_oficina)

online_query = ventas.query("canal == 'Online' and importe_neto > 500")
print("\nSelección con query:")
print(online_query)
print("query y la máscara coinciden:",online_query.equals(online_mayor_500))

ingenieria_remoto_senior = empleados[
    (empleados["departamento"] == "Ingenieria")
    & (empleados["modalidad"] == "Remoto")
    & (empleados["antiguedad_anios"] > 5)
].sort_values("salario_base", ascending=False)

print("\nIngeniería en remoto con más de cinco años:")
print(ingenieria_remoto_senior)

# Conservamos el índice porque contiene el identificador del empleado.
ingenieria_remoto_senior.to_csv("../outputs/ingenieria_remoto_senior.csv", encoding="utf-8")
print("\nArchivo guardado: ../outputs/ingenieria_remoto_senior.csv")


Pedidos Online de más de 500 euros:
     id_pedido       fecha  region    categoria          producto  unidades  \
7    PED-10193  2025-01-04  Centro      Deporte   Zapatilla Trail         5   
16   PED-10183  2025-01-09   Oeste        Hogar   Robot Aspirador         3   
22   PED-10130  2025-01-14    Este  Electronica        Tablet 10"         8   
23   PED-10149  2025-01-15    Este        Hogar   Robot Aspirador         3   
28   PED-10057  2025-01-18  Centro      Oficina  Teclado Mecanico         6   
..         ...         ...     ...          ...               ...       ...   
390  PED-10287  2025-09-10  Centro      Oficina  Silla Ergonomica         3   
393  PED-10310  2025-09-11     Sur  Electronica     Smartwatch S2         5   
399  PED-10128  2025-09-15  Centro  Electronica        Tablet 10"         3   
412  PED-10388  2025-09-27     Sur        Hogar   Cafetera Expres         3   
413  PED-10392  2025-09-27   Oeste  Electronica     Smartwatch S2         7   

     precio_uni

Con `query()` el filtro se lee más seguido, mientras que con corchetes vemos cada condición separada mediante `&`.


## Ejercicio 19 — Bucles sobre estructuras de datos

In [21]:
stock = 12
lote = 0

# Seguimos reponiendo mientras no alcancemos el objetivo de 60 unidades.
while stock < 60:
    lote = lote + 1
    stock = stock + 8
    print("Lote", lote, "-> stock:", stock)

regiones = ["Norte", "Sur", "Este", "Oeste", "Centro"]
print("\nRegiones en mayúsculas:")
for region in regiones:
    print(region.upper())

print("\nRegiones numeradas:")
for numero, region in enumerate(regiones, start=1):
    print(f"Región {numero}: {region}")

print("\nValor del stock por fila:")
for almacen, producto, unidades, coste in inventario:
    print(almacen, "|", producto, "|", round(unidades * coste, 2), "€")

print("\nTarifas de envío:")
for region, tarifa in tarifas_envio.items():
    print(f"{region}: {tarifa:.2f} €")

print("\nPrecios con recargo del 4 %:")
for precio in np_precios:
    print(round(precio * 1.04,2))


Lote 1 -> stock: 20
Lote 2 -> stock: 28
Lote 3 -> stock: 36
Lote 4 -> stock: 44
Lote 5 -> stock: 52
Lote 6 -> stock: 60

Regiones en mayúsculas:
NORTE
SUR
ESTE
OESTE
CENTRO

Regiones numeradas:
Región 1: Norte
Región 2: Sur
Región 3: Este
Región 4: Oeste
Región 5: Centro

Valor del stock por fila:
ALM-NORTE | Robot Aspirador | 6851.0 €
ALM-NORTE | Monitor 27 | 8236.0 €
ALM-SUR | Silla Ergonomica | 2146.8 €
ALM-SUR | Robot Aspirador | 1410.5 €
ALM-ESTE | Cafetera Expres | 5379.2 €
ALM-ESTE | Monitor 27 | 3550.0 €

Tarifas de envío:
Norte: 4.95 €
Este: 5.20 €
Oeste: 5.95 €
Centro: 3.90 €
Insular: 9.80 €

Precios con recargo del 4 %:
62.3
290.16
19.24
154.96
93.5
28.6
227.76
15.5
66.56
248.56


Usamos `.items()` para obtener la clave y el valor de cada entrada del diccionario, porque si lo recorremos directamente solo obtenemos las claves.

## Ejercicio 20 — Iteración sobre DataFrames

In [22]:
for etiqueta, fila in ventas.head(5).iterrows():
    print(fila["region"], "|", fila["producto"], "|", fila["importe_neto"], "€")

ventas["codigo_region"] = ventas["region"].str[:3].str.upper()
ventas["segmento_ticket"] = ventas["importe_neto"].apply(clasificar_ticket)
print("\nCódigos y segmentos:")
print(ventas[["region", "codigo_region", "importe_neto", "segmento_ticket"]].head())

print("\nPedidos por segmento:")
print(ventas["segmento_ticket"].value_counts())

# La división se aplica a todas las lecturas sin recorrerlas con un bucle.
sensores["eficiencia"] = (
    sensores["unidades_producidas"] / sensores["temperatura_c"]
).round(3)
print("\nEficiencia de las primeras lecturas:")
print(sensores[["unidades_producidas", "temperatura_c", "eficiencia"]].head())

resumen_por_region = {}
for region in ventas["region"].unique():
    filtrado = ventas[ventas["region"] == region]
    resumen_por_region[region] = round(filtrado["importe_neto"].sum(), 2)

print("\nResumen calculado con el bucle:")
print(resumen_por_region)

# Comparamos por región; el orden de aparición no cambia los importes.
resumen_groupby = facturacion_region.round(2).to_dict()
print("Resumen con groupby del ejercicio 15:")
print(resumen_groupby)
print("Las cifras coinciden:",resumen_por_region == resumen_groupby)

region_lider = max(resumen_por_region, key=resumen_por_region.get)
print("Región con mayor facturación:", region_lider)
print("Facturación de la región líder:", resumen_por_region[region_lider])


Centro | Tablet 10" | 350.4 €
Sur | Zapatilla Trail | 476.0 €
Oeste | Teclado Mecanico | 242.73 €
Oeste | Robot Aspirador | 1422.9 €
Sur | Mochila 30L | 256.0 €

Códigos y segmentos:
   region codigo_region  importe_neto segmento_ticket
0  Centro           CEN        350.40           Medio
1     Sur           SUR        476.00            Alto
2   Oeste           OES        242.73           Medio
3   Oeste           OES       1422.90            Alto
4     Sur           SUR        256.00           Medio

Pedidos por segmento:
segmento_ticket
Alto     167
Medio    152
Bajo     101
Name: count, dtype: int64

Eficiencia de las primeras lecturas:
   unidades_producidas  temperatura_c  eficiencia
0                  141          58.92       2.393
1                  153          71.72       2.133
2                  117          67.93       1.722
3                  105          73.10       1.436
4                   85          69.37       1.225

Resumen calculado con el bucle:
{'Centro': np.floa

Prefiero `groupby()` porque permite sumar las ventas por región con menos código y resulta más fácil de leer.
